File: kube/src/main/java/com/jacksonrakena/infrastructure/apps/Mixer.kt
package com.jacksonrakena.infrastructure.apps

import com.jacksonrakena.infrastructure.util.applyCommonConfiguration
import com.jacksonrakena.infrastructure.util.kube
import org.cdk8s.Chart
import org.cdk8s.ChartProps
import org.cdk8s.plus28.ContainerPort
import org.cdk8s.plus28.ContainerProps
import org.cdk8s.plus28.Deployment
import org.cdk8s.plus28.DeploymentExposeViaServiceOptions
import org.cdk8s.plus28.DeploymentProps
import org.cdk8s.plus28.DeploymentStrategy
import org.cdk8s.plus28.DockerConfigSecret
import org.cdk8s.plus28.EnvFrom
import org.cdk8s.plus28.HttpGetProbeOptions
import org.cdk8s.plus28.IConfigMap
import org.cdk8s.plus28.Probe
import org.cdk8s.plus28.ServicePort
import software.constructs.Construct
import kotlin.time.Duration.Companion.seconds

class Mixer(
    scope: Construct,
    id: String,
    configMap: IConfigMap,
    registrySecret: DockerConfigSecret,
    props: ChartProps? = null,
) : Chart(scope, id, props) {
    val deployment = Deployment(
        this,
        "deployment",
        DeploymentProps.builder()
            .replicas(1)
            .strategy(
                DeploymentStrategy.recreate()
            )
            .containers(
                listOf(
                    ContainerProps.builder()
                        .name("mixer-api")
                        .applyCommonConfiguration()
                        .image("ghcr.io/jacksonrakena/mixer:latest")
                        .ports(
                            listOf(
                                ContainerPort.builder().number(8080).build()
                            )
                        )
                        .envFrom(
                            listOf(
                                EnvFrom(configMap)
                            )
                        )
                        .liveness(
                            Probe.fromHttpGet(
                                "/actuator/health", HttpGetProbeOptions.builder().initialDelaySeconds(
                                    3.seconds.kube
                                ).periodSeconds(3.seconds.kube).port(8080).build()
                            )
                        )
                        .build()
                )
            )
            .dockerRegistryAuth(registrySecret)
            .build()
    )

    val service = deployment.exposeViaService(
        DeploymentExposeViaServiceOptions.builder()
            .ports(
                listOf(
                    ServicePort.builder().port(80).targetPort(8080).build()
                )
            ).build()
    )
}

File: src/apps/mixer.ts
import { Construct } from "constructs";
import { Chart, ChartProps, Duration } from "cdk8s";
import * as kplus from "cdk8s-plus-28";
import { applyCommonConfiguration } from "../util/container-utils";

export class Mixer extends Chart {
  public readonly service: kplus.Service;
  public readonly frontendService: kplus.Service;

  constructor(
    scope: Construct,
    id: string,
    configMap: kplus.IConfigMap,
    registrySecret: kplus.DockerConfigSecret,
    postgresService: kplus.Service,
    props?: ChartProps,
  ) {
    super(scope, id, props);

    const deployment = new kplus.Deployment(this, "backend", {
      replicas: 3,
      containers: [
        applyCommonConfiguration({
          name: "mixer-api",
          image: "ghcr.io/jacksonrakena/mixer:latest",
          ports: [{ number: 8080 }],
          envFrom: [new kplus.EnvFrom(configMap)],
          envVariables: {
            SPRING_DATASOURCE_URL: kplus.EnvValue.fromValue(
              `jdbc:postgresql://${postgresService.name}/mixer`,
            ),
            SPRING_PROFILES_ACTIVE: kplus.EnvValue.fromValue("prod"),
            SPRINGDOC_API_DOCS_ENABLED: kplus.EnvValue.fromValue("false"),
            SPRINGDOC_SWAGGER_UI_ENABLED: kplus.EnvValue.fromValue("false"),
            MIXER_REFRESH_FX_INITIAL: kplus.EnvValue.fromValue("-1"),
            MIXER_REFRESH_AGGREGATIONS_INITIAL: kplus.EnvValue.fromValue("-1"),
          },
          liveness: kplus.Probe.fromHttpGet("/actuator/health/liveness", {
            initialDelaySeconds: Duration.seconds(20),
            periodSeconds: Duration.seconds(10),
            port: 8080,
          }),
          readiness: kplus.Probe.fromHttpGet("/actuator/health/readiness", {
            initialDelaySeconds: Duration.seconds(20),
            periodSeconds: Duration.seconds(15),
            port: 8080,
          }),
        }),
      ],
      dockerRegistryAuth: registrySecret,
    });

    this.service = deployment.exposeViaService({
      ports: [{ port: 80, targetPort: 8080 }],
    });

    const frontendDeployment = new kplus.Deployment(this, "frontend", {
      replicas: 3,
      containers: [
        applyCommonConfiguration({
          name: "mixer-frontend",
          image: "ghcr.io/jacksonrakena/mixer:latest-frontend",
          ports: [{ number: 3000 }],
          envVariables: {
            API_BASE_URL: kplus.EnvValue.fromValue(
              "https://finance-api.rakena.com.au",
            ),
          },
        }),
      ],
      dockerRegistryAuth: registrySecret,
    });

    this.frontendService = frontendDeployment.exposeViaService({
      ports: [{ port: 80, targetPort: 3000 }],
    });
  }
}
